### DISCUSSION TITLE
Help with on connect event and parameters

### DISCUSSION THREAD

**Discussion by AdolfoBanchio** [miguelgrinberg/python-socketio#1253] [Q&A] [answered]
Hello, im working on a fastAPI application and i have managed to setup the connection of the socket and test it with postman.
this is my main.py
```python
app = FastAPI()
app.include_router(games_endpoints.router)
app.include_router(players_router)

origins = ["*"]
app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


@app.get("/")
async def root():
    return {"message": "La Cosa"}


sio = socketio.AsyncServer(cors_allowed_origins='*', async_mode='asgi')
socketio_app = socketio.ASGIApp(sio, app, socketio_path='/sockets')


@sio.event
def connect(sid, environ):
    print("connect ", sid)


@sio.on('message')
async def chat_message(sid, data):
    print("message ", data, "hola")
    await sio.emit('response', 'hola mundo')


@sio.on('join')
async def join(sid, data):
    print("join ", data)
    await sio.emit('response', 'te uniste'+data, room=sid)
@sio.event
def disconnect(sid):
    print('disconnect ', sid)


db.bind(provider="sqlite", filename=DATABASE_FILENAME, create_db=True)
db.generate_mapping(create_tables=True)
```

i want to in my connect event to save some parameters in the session of the sid that connected. This is how im connecting from postman to the socket

![image](https://github.com/miguelgrinberg/python-socketio/assets/78754770/037b2650-ee78-45be-a272-ddacdd14967f)
for testing i included the parameters needed in headers and as query parameters.

Now i dont know how to acces them correctyl in my on connect function so i can save them in the session using save_session()

**Comment by miguelgrinberg** [comment, marked as answer]
The connect event receives a dictionary with the request data, formatted according to the WSGI specification. It's unclear from your screenshot if you intend to pass the player and game ids as query string arguments or as custom headers, but all of this information is in the `environ` dictionary:

- for the query string, use `environ['QUERY_STRING']`
- for the headers use `environ['HTTP_PLAYER_ID']` and `environ['HTTP_GAME_ID']`

**Reply by AdolfoBanchio** [reply]
Sorry for my mistake. 'for the headers use environ['HTTP_PLAYER_ID'] and environ['HTTP_GAME_ID']' this was what i was looking for. Thanks!!

**Reply by AdolfoBanchio** [reply]
If i want to use query parameters instead. The only way to acces them in server side is by parsing the string saved in ```environ['QUERY_STRING']```  ?

**Reply by miguelgrinberg** [reply]
Yes. You can use the [parse_qs()](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.parse_qs) function from the Python standard library to convert the string to a dictionary.

**Reply by AdolfoBanchio** [reply]
Thanks! and another question. Im creating a cards game, so for each player i create a room with name p+{playerid}, and all players with same game id i also add them to a room with name g+{gameid}. 
When the game finishes all the connections are closed. I would like to know if i can close those rooms when disconnected or the rooms are deleted automatically.

**Reply by miguelgrinberg** [reply]
Rooms are deleted when they become empty. And when a client disconnects, it is removed from all the rooms it was in.

In other words, as long as your clients eventually go away, there shouldn't be any lingering rooms. Of course you are welcome to close rooms manually if you prefer, and then any participants are removed, even if they are still connected (like maybe playing another game). There is a `close_room()` function for this.

**Reply by AdolfoBanchio** [reply]
Okay, thank you very much! Everything was really helpful
